pennylane/tape/qscript.py

```diff
@@ -438,6 +447,20 @@ def _update_observables(self):
                     self._obs_sharing_wires.append(m.obs)
                     self._obs_sharing_wires_id.append(i)
 
+    def _update_measured_wires(self):
+        """Updates the measured wires straightforwardly.
+
+        Sets:
+            _measured_wires (int): The wires measured in this QuantumScript
+        """
+        wires = []
+
+        for m in self.measurements:
+            if m.obs is not None:
+                wires.append(m.obs.wires)
+
+        self._measured_wires = Wires.all_wires([*wires, self._measured_wires])
```

Comment:
why do you do a union with `self._measured_wires`? this gets messy because QuantumScript is mutable, so it might be created measuring wires `[0, 1]`, then later be updated to measure `[2]`. With this code, if someone calls `qscript.measured_wires`, they'll get `[0, 1, 2]` but should probably only get `[2]`.

The challenge here is really deciding when to re-compute this attribute, because we don't exactly have a listener checking if `self.measurements` has been changed.

Reply:
Thanks for raising this. It seems to me to be a use case issue that would be resolved by a best practice. Is there documentation for using QuantumScript that specifies when to mutate it versus when to instantiate a new one?

I would probably say that the user is taking things into their own hands when they mutate a QuantumScript. The question pertaining to this ticket seems to be whether the measured wires property remembers all the wires that are measured by the QuantumScript over time or not. Of course when dealing with a cached property, this is a pertinent question. I am open to whatever you think is the best behaviour.

Reply:
haha you're touching on questions I'm scared to answer! some thoughts:
- I agree that the user is taking things into their own hands when mutating. On the one hand, QuantumScript is a rather internal detail (hence why it's in the "Internal" section in the docs sidebar 😄) so typical users shouldn't be near it. On the other hand, it happens anyway. So I would strive to make it as clear as possible to the user that they are in fact mutating an existing object, and they know what the consequences are. Once informed, PennyLane has done its job and they can do what they like!
- I don't think PL devs have ever thought about a QuantumScript as something that evolves with time, so I'm inclined to say that if a user _measured_ some wires, changed observables, then _measured_ again, they should probably make a new tape. Tapes often undergo various transformations, but I think usually before execution, and not in this way.

It's not totally clear to me what's best, so I will let you decide! Please make sure to document the assumption that you make, and I'm fine either way

Reply:
Thanks for your thoughts. In this case I may just leave it as-is and assume as you have said that the QuantumScript won't evolve with time.

Reply:
I've made this change in a recent commit.